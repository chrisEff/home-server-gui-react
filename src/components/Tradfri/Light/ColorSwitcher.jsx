'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'
import {setTradfriBulbColor} from '../../../actions'

const ColorSwitcher = ({deviceId, bulbType, color, onChange}) => {
	if (bulbType === 'rgb') {
		return (
			<div className='color-switcher'>
				<div className={color === 'red'     ? 'color red active'     : 'color red'}     onClick={() => onChange(deviceId, 'red')}><div></div></div>
				<div className={color === 'green'   ? 'color green active'   : 'color green'}   onClick={() => onChange(deviceId, 'green')}><div></div></div>
				<div className={color === 'blue'    ? 'color blue active'    : 'color blue'}    onClick={() => onChange(deviceId, 'blue')}><div></div></div>
				<div className={color === 'yellow'  ? 'color yellow active'  : 'color yellow'}  onClick={() => onChange(deviceId, 'yellow')}><div></div></div>
				<div className={color === 'pink'    ? 'color pink active'    : 'color pink'}    onClick={() => onChange(deviceId, 'pink')}><div></div></div>
				<div className={color === 'purple'  ? 'color purple active'  : 'color purple'}  onClick={() => onChange(deviceId, 'purple')}><div></div></div>
				<div className={color === 'warm'    ? 'color warm active'    : 'color warm'}    onClick={() => onChange(deviceId, 'warm')}><div></div></div>
				<div className={color === 'neutral' ? 'color neutral active' : 'color neutral'} onClick={() => onChange(deviceId, 'neutral')}><div></div></div>
				<div className={color === 'cold'    ? 'color cold active'    : 'color cold'}    onClick={() => onChange(deviceId, 'cold')}><div></div></div>
			</div>
		)
	}
	if (bulbType === 'white-spectrum') {
		return (
			<div className='color-switcher spectrum'>
				<div className={color === 'warm'    ? 'color warm active'    : 'color warm'}    onClick={() => onChange(deviceId, 'warm')}><div></div></div>
				<div className={color === 'neutral' ? 'color neutral active' : 'color neutral'} onClick={() => onChange(deviceId, 'neutral')}><div></div></div>
				<div className={color === 'cold'    ? 'color cold active'    : 'color cold'}    onClick={() => onChange(deviceId, 'cold')}><div></div></div>
			</div>
		)
	}
	if (bulbType === 'white') {
		return (
			<div className='color-switcher white'>
				<div className='color neutral active'><div></div></div>
			</div>
		)
	}
	return <div/>
}

ColorSwitcher.propTypes = {
	deviceId: PropTypes.number.isRequired,
	bulbType: PropTypes.string.isRequired,
	color: PropTypes.string,
	onchange: PropTypes.func,
}

const mapStateToProps = (state, ownProps) => ({
})
const mapDispatchToProps = dispatch => ({
	onChange: (id, color) => dispatch(setTradfriBulbColor(id, color)),
})

export default connect(mapStateToProps, mapDispatchToProps)(ColorSwitcher)
export {ColorSwitcher}
