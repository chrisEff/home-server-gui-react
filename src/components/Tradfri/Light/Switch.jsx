'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'
import {setTradfriiDeviceState} from '../../../actions'

let Switch = ({deviceId, state, onChange}) =>
	<div className='bulb' onClick={() => onChange(deviceId, state ? 0 : 1)}>ðŸ’¡</div>

Switch.propTypes = {
	deviceId: PropTypes.number.isRequired,
	state: PropTypes.number.isRequired,
	onchange: PropTypes.func,
}

const mapStateToProps = (state, ownProps) => ({
})
const mapDispatchToProps = dispatch => ({
	onChange: (id, state) => dispatch(setTradfriiDeviceState(id, state)),
})

export default connect(mapStateToProps, mapDispatchToProps)(Switch)
export {Switch}
