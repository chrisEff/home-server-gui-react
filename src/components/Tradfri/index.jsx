'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'

import LightGroup from './LightGroup'
import {loadTradfriDevices, loadTradfriGroups} from '../../actions'

class Tradfri extends React.Component {

	static propTypes = {
		title: PropTypes.string,
	}

	async componentDidMount () {
		this.props.onLoad()
	}

	render () {
		return (
			<div id='tradfri'>
				<h2>{this.props.title}</h2>
				{this.props.tradfriGroups.map(group => <LightGroup key={group.name} id={group.id} name={group.name} deviceIds={group.deviceIds}/>)}
			</div>
		)
	}
}

const mapStateToProps = state => ({
	tradfriGroups: state.tradfriGroups,
	tradfriDevices: state.tradfriDevices,
})

const mapDispatchToProps = dispatch => ({
	onLoad: () => {
		dispatch(loadTradfriGroups())
		dispatch(loadTradfriDevices())
	},
})

export default connect(mapStateToProps, mapDispatchToProps)(Tradfri)
export {Tradfri}
