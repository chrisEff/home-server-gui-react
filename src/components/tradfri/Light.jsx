'use strict'

const api = require('../../../config').api
api.url = `${api.protocol}://${api.host}:${api.port}`

const React = require('react')
const PropTypes = require('prop-types')

const LightTitle = require('./LightTitle')
const BrightnessSlider = require('./BrightnessSlider')
const ColorSwitcher = require('./ColorSwitcher')
const LightSwitch = require('./LightSwitch')

class Light extends React.Component {

	static propTypes = {
		bulb: PropTypes.object.isRequired,
	}

	constructor (props) {
		super(props)
		this.state = {bulb: props.bulb}
	}

	render () {
		const bulb = this.state.bulb
		const className = bulb.state ? 'light on' : 'light'
		return (
			<div className={className + ' ' + bulb.color}>
				<LightTitle title={bulb.name} onChange={this.updateName} />

				<LightSwitch state={bulb.state} onchange={this.updateState}/>
				<BrightnessSlider brightness={bulb.brightness} onchange={this.updateBrightness}/>
				<ColorSwitcher bulbType={bulb.bulbType} color={bulb.color} onchange={this.updateColor}/>
			</div>
		)
	}

	updateName = async (name) => {
		const bulb = this.state.bulb
		await fetch(`${api.url}/tradfri/device/${bulb.id}/name/${name}?key=${api.key}`, {method: 'PUT'})

		bulb.name = name
		this.setState({bulb})
	}

	updateState = async (state) => {
		const bulb = this.state.bulb
		await fetch(`${api.url}/tradfri/device/${bulb.id}/state/${state}?key=${api.key}`, {method: 'PUT'})

		bulb.state = state
		this.setState({bulb})
	}

	updateBrightness = async (brightness) => {
		const bulb = this.state.bulb
		await fetch(`${api.url}/tradfri/device/${bulb.id}/brightness/${brightness}?key=${api.key}`, {method: 'PUT'})

		bulb.brightness = brightness
		this.setState({bulb})
	}

	updateColor = async (color) => {
		const bulb = this.state.bulb
		await fetch(`${api.url}/tradfri/device/${bulb.id}/color/${color}?key=${api.key}`, {method: 'PUT'})

		bulb.color = color
		this.setState({bulb})
	}
}

module.exports = Light
