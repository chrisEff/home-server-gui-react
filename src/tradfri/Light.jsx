'use strict'

const config = require('../../config')

const React = require('react')
const BrightnessSlider = require('./BrightnessSlider')
const ColorSwitcher = require('./ColorSwitcher')
const LightSwitch = require('./LightSwitch')

class Light extends React.Component {
	constructor (props) {
		super(props)
		this.state = {bulb: props.bulb}
	}

	render () {
		const bulb = this.state.bulb;
		const className = bulb.state ? 'light on' : 'light'
		return (
			<div className={className + ' ' + bulb.color}>
				<span className={bulb.name.length > 15 ? 'name long' : 'name'}>{bulb.name}</span>

				<LightSwitch state={bulb.state} onchange={this.updateState}/>
				<BrightnessSlider brightness={bulb.brightness} onchange={this.updateBrightness}/>
				<ColorSwitcher bulbType={bulb.bulbType} color={bulb.color} onchange={this.updateColor}/>
			</div>
		)
	}

	updateState = async (state) => {
		const bulb = this.state.bulb
		await fetch(
			`${config.api.protocol}://${config.api.host}:${config.api.port}/tradfri/device/${bulb.id}/state/${state}?key=${config.api.key}`,
			{method: 'PUT'}
		)

		bulb.state = state
		this.setState({bulb})
	}

	updateBrightness = async (brightness) => {
		const bulb = this.state.bulb
		await fetch(
			`${config.api.protocol}://${config.api.host}:${config.api.port}/tradfri/device/${bulb.id}/brightness/${brightness}?key=${config.api.key}`,
			{method: 'PUT'}
		)

		bulb.brightness = brightness
		this.setState({bulb})
	}

	updateColor = async (color) => {
		const bulb = this.state.bulb
		await fetch(
			`${config.api.protocol}://${config.api.host}:${config.api.port}/tradfri/device/${bulb.id}/color/${color}?key=${config.api.key}`,
			{method: 'PUT'}
		)

		bulb.color = color
		this.setState({bulb})
	}
}

module.exports = Light
