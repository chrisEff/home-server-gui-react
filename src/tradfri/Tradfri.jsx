'use strict'

const config = require('../../config')
const LightGroup = require('./LightGroup')
const ErrorMessage = require('../ErrorMessage')

class Tradfri extends React.Component {
	constructor (props) {
		super(props)
		this.state = {
			groups: [],
			errorMsg: null,
		}
	}

	async componentDidMount () {
		try {
			const groups = await (await fetch(`${config.api.protocol}://${config.api.host}:${config.api.port}/tradfri/group?key=${config.api.key}`)).json()
			const devices = await (await fetch(`${config.api.protocol}://${config.api.host}:${config.api.port}/tradfri/device?key=${config.api.key}`)).json()

			groups.forEach(group => {
				group.devices = []
				group.deviceIds.forEach(deviceId => {
					const device = devices.find(device => device.id === deviceId)
					group.devices.push(device)
				})
				group.devices.sort((a, b) => {
					return a.name > b.name ? 1 : b.name < a.name ? -1 : 0;
				})
			})
			groups.sort((a, b) => {
				return a.name > b.name ? 1 : b.name < a.name ? -1 : 0;
			})

			this.setState({
				groups
			})
		} catch (e) {
			this.setState({errorMsg: e.message})
		}
	}

	render () {
		return (
			<div>
				<h2>Licht</h2>
				{this.state.errorMsg && <ErrorMessage message={this.state.errorMsg}/>}
				{this.state.groups.map(group => <LightGroup key={group.name} group={group}/>)}
			</div>
		)
	}
}

module.exports = Tradfri
