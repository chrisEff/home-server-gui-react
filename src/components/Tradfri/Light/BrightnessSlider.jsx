'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { StyleSheet, css } from 'aphrodite'

import { setDeviceBrightness } from '@/actions/tradfri'

const BrightnessSlider = ({ deviceId, brightness, onChange }) =>
	<input
		className={css([styles.brightnessSlider])}
		type="range"
		min="0"
		max="254"
		defaultValue={brightness}
		onMouseUp={(e) => onChange(deviceId, parseInt(e.target.value))}
	/>

BrightnessSlider.propTypes = {
	deviceId: PropTypes.number.isRequired,
	brightness: PropTypes.number.isRequired,
	onchange: PropTypes.func,
}

const styles = StyleSheet.create({
	brightnessSlider: {
		cursor: 'pointer',
		margin: '20px 0',
		width: '120px',
	},
})

const mapStateToProps = (state, ownProps) => ({})
const mapDispatchToProps = (dispatch) => ({
	onChange: (id, brightness) => dispatch(setDeviceBrightness(id, brightness)),
})

export default connect(mapStateToProps, mapDispatchToProps)(BrightnessSlider)
export { BrightnessSlider }
