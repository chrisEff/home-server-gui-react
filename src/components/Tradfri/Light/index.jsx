'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'

import Title from './Title'
import BrightnessSlider from './BrightnessSlider'
import ColorSwitcher from './ColorSwitcher'
import Switch from './Switch'

const Light = ({bulb}) => {
	const className = bulb.state ? 'light on' : 'light'
	return (
		<div className={className + ' ' + bulb.color}>
			<Title deviceid={bulb.id} title={bulb.name} />
			<Switch deviceid={bulb.id} state={bulb.state} />
			<BrightnessSlider deviceid={bulb.id} brightness={bulb.brightness} />
			<ColorSwitcher deviceid={bulb.id} bulbType={bulb.bulbType} color={bulb.color} />
		</div>
	)
}

Light.propTypes = {
	id: PropTypes.number,
	bulb: PropTypes.object.isRequired,
}

const mapStateToProps = (state, ownProps) => ({
	bulb: state.tradfriDevices.find(device => device.id === ownProps.id),
})
const mapDispatchToProps = dispatch => ({
})

export default connect(mapStateToProps, mapDispatchToProps)(Light)
export {Light}
