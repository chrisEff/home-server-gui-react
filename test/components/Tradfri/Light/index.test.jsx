'use strict'

import React from 'react'
import { Light } from '@/components/Tradfri/Light'

import { shallow } from 'enzyme'

describe('Light', () => {
	it('should render correctly', () => {
		const wrapper = shallow(
			<Light
				bulb={{
					brightness: 254,
					bulbType: 'white-spectrum',
					color: 'neutral',
					firmware: '1.2.217',
					id: 65537,
					model: 'TRADFRI bulb E27 WS opal 980lm',
					name: 'Decke',
					state: 0,
					type: 'bulb',
				}}
			/>,
		)
		expect(wrapper).toMatchSnapshot()
	})
})
