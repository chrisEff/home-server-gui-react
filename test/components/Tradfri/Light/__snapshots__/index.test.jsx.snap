// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Light should render correctly 1`] = `
<div
  className="light neutraål"
>
  <Connect(Title)
    deviceid={65537}
    title="Decke"
  />
  <Connect(Switch)
    deviceid={65537}
    state={0}
  />
  <Connect(BrightnessSlider)
    brightness={254}
    deviceid={65537}
  />
  <Connect(ColorSwitcher)
    bulbType="white-spectrum"
    color="neutraål"
    deviceid={65537}
  />
</div>
`;
